### DISCUSSION TITLE
How to add this data source using grafanas http api when an api key is required for authentication

### DISCUSSION THREAD

**Discussion by LMH01** [grafana/grafana-infinity-datasource#801] [Q&A] [answered]
Hello everyone,

I would like to know how I can add this data source using the http api, when my authentication method is an api key that is submitted in the http header.

I got it working so far with this json data, but the problem is that I don't know what the field of the accepted api key is. I tried `apiKeyValue` but when I submitted the request to the server with that value it was not set:

```json
{
   "access":"proxy",
   "basicAuth":false,
   "basicAuthUser":"",
   "database":"",
   "isDefault":true,
   "jsonData":{
      "allowedHosts":[
         "http://localhost:11511"
      ],
      "apiKeyKey":"api_key",
      "apiKeyValue": "key_value", # <-- does not work
      "auth_method":"apiKey",
      "customHealthCheckEnabled":false,
      "customHealthCheckUrl":"",
      "global_queries":[
         
      ],
      "oauthPassThru":false
   },
   "name":"yesoreyeram-infinity-datasource",
   "orgId":1,
   "readOnly":false,
   "secureJsonFields":{
      "apiKeyValue":true
   },
   "type":"yesoreyeram-infinity-datasource",
   "typeLogoUrl":"public/plugins/yesoreyeram-infinity-datasource/img/icon.svg",
   "uid":"my-unique-id",
   "url":"",
   "user":"",
   "version":3,
   "withCredentials":false
}
```

When this json data was submitted this was the result in grafana:

![image](https://github.com/grafana/grafana-infinity-datasource/assets/35499705/47a56f12-f734-461e-a20c-f61d9c5a0184)

So all values where properly set but the api key value was not. I tried looking at the documentation but I was not able to find anything.

Does anyone know how I can submit that api key value in the api request to the grafana intance to automatically create this data source where the api key value is set? Is this even possible?

**Comment by yesoreyeram** [comment, marked as answer]
`apiKeyValue` need to go inside `secureJsonData`, not inside `jsonData`. When you need help with the structure of the infinity config, use the **provisioning script** in the datasource config as shown below. ( When storing the sensitive data, you will use `secureJsonData`. When retrieving sensitive data, you will see `secureJsonFields` in your grafana api response )

<img width="1516" alt="image" src="https://github.com/grafana/grafana-infinity-datasource/assets/153843/983016e4-9e75-4909-bcea-570c9bd00651">

**Reply by LMH01** [reply]
Thanks for your answer, this works!
